src/odemis/acq/acqmng.py

```diff
@@ -83,6 +85,202 @@ def acquire(streams, settings_obs=None):
+                        self._main_future.set_end_time(time.time() + remaining_t)
+
+                    try:
+                        # acquire this single stream, and get the data
+                        self._single_acqui_f = acquire([stream], self._settings_obs)
+                        data, exp = self._single_acqui_f.result()
+
+                        # check if cancellation happened while the acquiring future is working
+                        if self._future_state == CANCELLED:
+                            raise CancelledError()
+
+                        # check on the acquired data
+                        if not data:
+                            logging.warning("The acquired data array for stream %s is empty", stream)
+                        else:
+                            logging.info("The acquisition for stream %s is done", stream)
+
+                    except CancelledError:
+                        raise
+                    except Exception as e:
+                        logging.exception("The acquisition failed at the %s-th zlevel of the stream %s, because %s" %(i+1, stream, e))
+                        # TODO handle zstack assembling in case of error 
+                        return acquired_data, e
+                    
+                    zstack.append(data[0])
+                    # update the remaining time
+                    remaining_t -= stream.estimateAcquisitionTime()
+                    self._main_future.set_end_time(time.time() + remaining_t)
+
+                zcube = assembleZCube(zstack, self._zlevels[stream])
+                acquired_data.append(zcube)
+
+        # state that the future has finished
+        self._future_lock.acquire() 
+        self._future_state = FINISHED
+        self._future_lock.release()
+
+        return acquired_data, exp
+
+
+def estimateZStackAcquisitionTime(streams, zlevels):
+    """
+    Estimates the total time for zstack acquisition including the movements time 
+        of the zsteps 
+    streams(list of Stream): the streams to be acquired 
+    zlevels(dict Stream -> list of float): the zlevels dictionary containing
+        the streams along with the list of z positions 
+    return (float): total estimated time of acquisition of streams and zsteps 
+    """
+    acq_time = 0
+    for s in streams:
+        if s in zlevels.keys():
+            acq_time += s.estimateAcquisitionTime() * len(zlevels[s])
+        else:
+            acq_time += s.estimateAcquisitionTime() 
+    # find the zstep, if applicable
+    z_pos = next((l for l in zlevels.values() if len(l) > 1), None)
+    if z_pos:
+        zstep = abs(z_pos[0]-z_pos[1])
+        # get the focuser of any of the FM streams
+        focuser = next((s.focuser for s in streams if isinstance(s, FluoStream)), None)
+        tot_zstep_time = (len(z_pos) - 1) * len(zlevels) * \
+            guessActuatorMoveDuration(focuser, axis="z", distance=zstep)
+        acq_time += tot_zstep_time
```

Comment:
This doesn't match exactly the actual acquisition code now we use a different focuser for each stream.

In practice, it's probably possible to even simplify the code. Something that would follow more the structure of the acquisition code:
```
for s in streams:
   zs = zlevels.get(s, [0])  # concider that streams without zlevels are acquired at a single z
   acq_time += s.estimateAcquisitionTime() * len(zs)
   if len(zs) > 1 and s.focuser:
       zstep = abs(zs[0] - zs[1])
       tot_zstep_time = (len(zs) - 1) *  guessActuatorMoveDuration(s.focuser, axis="z", distance=zstep)
       acq_time += tot_zstep_time
```